### PR TITLE
Store and yank gem contents in background jobs

### PR DESCRIPTION
**Pull request by martinemde**
Code complete. Needs a few things cleaned up like TODO comments.

### CODE REVIEW

**Comment by simi** [comment]
I have still my final comment in progress on the RFC, which wasn't merged. But do we really need to keep this inside the main app? Is there any reason why to not start this as an separated app? 🤔

**Comment by simi** [comment]
Can you please briefly explain how the gem content is stored? Is it stored unpacked to "storage" keeping the structure as in gem?

**Comment by martinemde** [comment]
@simi I think that would be a fine solution. There's almost no interaction between this system an the main app. Just triggering the jobs on push/yank and providing the information about what files are available and where. We could extract this quite easily.

I think this just comes down to infrastructure. If it's easy enough to make a new app with new infra, then I'm for it. One thing I discovered while working on this is that malicious files (tarbombs) could be used to DoS the extraction process. I would feel better if that risk could be confined to instances that have no ability to interrupt production.

@indirect Thoughts on above?

File: lib/gem_inclusions/storage.rb
Comment (martinemde) [inline lib/gem_inclusions/storage.rb:9]: @simi The easiest way to explain how content is stored is to point at this snippet here.

Files are stored at the FILE_KEY at a hash of their content.
Specs are stored as the .gemspec file itself (this may not be quite complete, still adding tests)
Path keys store the actual file path and contain the hash, so you can get to a file contents with two S3 gets.
Hashes of all files in a gem version are stored for efficiency to allow yanking to not need to read every file in every gem on yank.

**Comment by jchestershopify** [comment]
One request that's come from security researchers and academics is for us to retain yanked gem contents, so that malware gems can be analysed. From my reading of this code we just delete the files; is there some way we could move them to a private bucket so they can be shared with verified researchers?

**Comment by martinemde** [comment]
@jchestershopify That seems reasonable. Maybe versioning the S3 bucket would be a simple way to get this feature because I don't think it would otherwise impact cost if we store only one version of all the files. I think the only impact is that we would pay for 'deleted' content. Providing access to the information to select researchers could be done through some other mechanism then.

**Comment by jchestershopify** [comment]
I'm glad to hear it's already versioned in the originating bucket, that allays my concerns.

**Comment by simi** [comment]
@martinemde I left a brief comment on RFC itself regarding our discussion. Feel free to ping me once this is ready for another round of review. Currently (AFAIK) this is awaiting code changes to simplify the implementation or even send some improvements to RubyGems (for example https://github.com/rubygems/rubygems/pull/6390). :pray: :muscle:

File: app/jobs/ingest_version_contents_job.rb
Comment (segiddins) [inline app/jobs/ingest_version_contents_job.rb:6]: retry_on accepts multiple exceptions

File: app/jobs/yank_version_contents_job.rb
Comment (segiddins) [inline app/jobs/yank_version_contents_job.rb:2]: same as above, set job adapter

File: app/jobs/ingest_version_contents_job.rb
Comment (segiddins) [inline app/jobs/ingest_version_contents_job.rb:20]: instead of doing this in `perform`, add a `rescue_from(StandardError)` and do the the log / re-raise in there

File: app/jobs/ingest_version_contents_job.rb
Comment (segiddins) [inline app/jobs/ingest_version_contents_job.rb:11]: can also pass the version directly, active job will serialize it as a global ID

File: lib/gem_content.rb
Comment (segiddins) [inline lib/gem_content.rb:14]: should this use a different S3 bucket from where we store the full gems?

Comment (segiddins) [reply lib/gem_content.rb:14]: I think that would make it easier to attribute costs, potentially set different versioning policies, etc

File: lib/gem_content.rb
Comment (segiddins) [inline lib/gem_content.rb:35]: does it maybe make sense to make this class enumerable and allow lazily enumerating keys?

File: lib/gem_content/package.rb
Comment (segiddins) [inline lib/gem_content/package.rb:26]: do we also want to be saving the gemspec & whatever the other tar in a `.gem` is?

File: lib/gem_content/package.rb
Comment (segiddins) [inline lib/gem_content/package.rb:18]: can we report the gem full name and entry data as context in the call to `Rails.error.report` ?

File: lib/rubygem_fs.rb
Comment (segiddins) [inline lib/rubygem_fs.rb:108]: should this also bubble the error up? or `Rails.error.report` it?

File: lib/version_manifest.rb
Comment (segiddins) [inline lib/version_manifest.rb:9]: why would the version match the gem name pattern?

File: lib/gem_content/package.rb
Comment (martinemde) [reply lib/gem_content/package.rb:26]: I save the spec from Gem::Package#spec.to_ruby. The other file is the checksum that verifies the other two files are not corrupt.

File: lib/rubygem_fs.rb
Comment (martinemde) [reply lib/rubygem_fs.rb:108]: This only happens in tests. It would be normal for metadata to be missing if none was saved (like when gems are saved). I can't imagine the parse error actually happening but it seemed harmless.

File: lib/version_manifest.rb
Comment (martinemde) [reply lib/version_manifest.rb:9]: We use the same name pattern for both rubygem names and [version numbers and platforms (AKA `slug`, which is what I'm expecting here)](https://github.com/rubygems/rubygems.org/blob/master/app/models/version.rb#L23-L24). This is more of a sanity check to make sure you don't pass in the object since it's just called `version`. I didn't like naming it `version_slug` or something longer since it just made this whole class uglier.

**Comment by martinemde** [comment]
Updated with a mostly complete version. 

Not enqueueing jobs yet. I want to wait for that until we can run a few manually. I have enqueuing in another branch. I could add it here if we want to put it behind an ENV?

Need to confirm how to load the bucket name for content. I put some boilerplate. Do we have a better way?

**Review by segiddins — APPROVED** [review]

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:116]: S3 changes metadata to string => string, so I thought it was easier to convert explicitly so it's always consistent. It means metadata is == directly from S3 after symbolize_keys.

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:98]: Oh good point. Will do

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:49]: I know hash is a special method but I don't think it causes any problems in this case. I never hash Entries by the object itself. I could use `hex` (which npmjs.com uses) in this place.

The idea was to separate the hashing for file duplication from the checksum algorithm by using two different access methods depending on what I needed. (Hopefully I was consistent enough)

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:39]: Maybe. I intended to gracefully return nils when there is no metadata, but I think that's handled by the caller and it makes this method too weak. I'll change it.

File: app/models/pusher.rb
Comment (martinemde) [reply app/models/pusher.rb:129]: Only problem would be if adding that many jobs would burden the queue. It would be good to have this live and run a few manually before every gem push triggers a job. At the very least we want to make sure they are writing to the correct bucket and not doing anything dastardly before we run a few hundred or thousand jobs.

File: lib/gem_package_enumerator.rb
Comment (martinemde) [reply lib/gem_package_enumerator.rb:17]: Yes. I haven't upstreamed this yet, so this doesn't exist. I have it in a branch in rubygems. I'll make a comment about that.

File: app/models/gem_content.rb
Comment (martinemde) [reply app/models/gem_content.rb:44]: Ah yes, meant to write a test that failed before uncommenting this code, then forgot.

File: app/models/pusher.rb
Comment (indirect) [reply app/models/pusher.rb:129]: Now that LaunchDarkly support is merged, maybe you can use a feature flag to control whether this runs or not?

File: app/models/gem_content_entry.rb
Comment (indirect) [reply app/models/gem_content_entry.rb:49]: I would suggest that the method be named something besides `hash` if it isn't designed specifically to replace the method that Ruby uses for object keying in hashes etc. 👍🏻

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:98]: Unfortunately: 

```ruby
irb> MIME::Types['application/json'].first.binary?
=> true
```

Comment (segiddins) [reply app/models/gem_content_entry.rb:98]: gotcha. interesting. https://github.com/mime-types/mime-types-data/blob/7109b079da2c87924066c5c3e7eb9b65ffcec9e8/types/text.yaml#L250-L262

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb]: Updated.

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb]: I decided to have this raise InvalidMetadata if it comes in with keys missing. This would indicate corrupted data on S3 and would need to be fixed by reuploading.

File: lib/gem_package_enumerator.rb
Comment (martinemde) [reply lib/gem_package_enumerator.rb:19]: Removed. I'll add back in when I can make the change to rubygems and it gets accepted.

File: lib/gem_content.rb
Comment (martinemde) [reply lib/gem_content.rb]: Yes, added.

File: lib/gem_content.rb
Comment (martinemde) [reply lib/gem_content.rb]: We could add this when it's needed. Should be easy to add.

File: lib/version_manifest.rb
Comment (martinemde) [reply lib/version_manifest.rb]: I updated this to use number/platform because that saves some complexity in a real API.

File: app/models/gem_content.rb
Comment (martinemde) [reply app/models/gem_content.rb]: I removed the code. I think it actually costs more to check if the file exists first.

File: app/models/pusher.rb
Comment (martinemde) [reply app/models/pusher.rb]: Added

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb]: I committed a solution that adds some extended checking.

File: app/jobs/store_version_contents_job.rb
Comment (segiddins) [inline app/jobs/store_version_contents_job.rb:18]:
```suggestion
    logger.error "Storing gem contents for #{version} failed", error
```

should take advantage of the newly-merged semantic_logger

File: app/jobs/store_version_contents_job.rb
Comment (segiddins) [inline app/jobs/store_version_contents_job.rb:29]: no need to use `Rails.logger`, every subclass of `ApplicationJob` has its own logger :)

File: app/models/gem_content.rb
Comment (segiddins) [inline app/models/gem_content.rb:34]: commented code

File: app/models/pusher.rb
Comment (segiddins) [inline app/models/pusher.rb:135]: nit: let's add an `ld_context` method to the `user` model, and have it return something like:

```ruby
LaunchDarkly::LDContext.create({
    key: "user-id-#{id}",
    kind: "user",
    name:
})
```

just because the `display_id` can change over time, and we want the context key to be stable

**Review by segiddins — APPROVED** [review]

File: app/models/gem_content.rb
Comment (martinemde) [reply app/models/gem_content.rb:34]: haha, I swear I removed this. I'll try again.

File: app/models/pusher.rb
Comment (martinemde) [reply app/models/pusher.rb]: updated.

File: app/models/gem_content_entry.rb
Comment (simi) [inline app/models/gem_content_entry.rb:43]: Why `hex`? Wouldn't be `hash` better?

Comment (martinemde) [reply app/models/gem_content_entry.rb:43]: I went back and forth on this a bunch. I don't know anymore.

It's only internal. It's saved as `sha256` in metadata so we can change hex any time. I don't like hex as much as hash, but hash has special meaning. 🤷

File: test/unit/gemcutter/middleware/hostess_test.rb
Comment (simi) [inline test/unit/gemcutter/middleware/hostess_test.rb:8]: Any info why this is needed?

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb:43]: [`fingerprint`](https://en.wikipedia.org/wiki/Fingerprint_(computing)) is maybe the most accurate. The only reason I don't use `sha256` everywhere is that I want to have a non algorithm-coupled shorthand that could stay constant no matter how we fingerprint data to identify it. It's a unique identifier for the content separate from the file identity.

Comment (simi) [reply app/models/gem_content_entry.rb:43]: Yup, clear. Let's use `fingerprint` then.

File: lib/gem_package_enumerator.rb
Comment (simi) [inline lib/gem_package_enumerator.rb:10]: What about to keep the most interesting part (laziness) of this custom enumerator to be part of the name like `GemPackageLazyEnumerator`? IMHO the best would be to keep it as `Gem:Package::LazyEnumerator`. :thinking: 

Also is the comment above still proper? If I understand it well, there is no `Gem::Package#each_file` method anymore.

File: .rubocop.yml
Comment (simi) [inline .rubocop.yml:13]: :thinking: Any idea why this change is needed in here?

File: test/unit/gemcutter/middleware/hostess_test.rb
Comment (martinemde) [reply test/unit/gemcutter/middleware/hostess_test.rb:8]: I had a trouble with it trying to use the S3 mock instead of Local. It was transient, based on another change, but I didn't like that this directly tests an object that uses RubygemFs but didn't explicitly make sure it was mocking before each test.

File: lib/gem_package_enumerator.rb
Comment (martinemde) [reply lib/gem_package_enumerator.rb:10]: `Gem::Package::LazyEnumerator` works, but ultimately I want to add `each_file` to `Gem::Package`, so this naming discussion might be moot. This class would disappear if I had each_file, but I didn't want to delay this PR while I worked on it.

File: app/jobs/store_version_contents_job.rb
Comment (simi) [inline app/jobs/store_version_contents_job.rb:31]: Does this load whole gem into memory? There are even 450MB gems - https://rubygems.org/gems/emojidex-rasters/versions/1.0.34.

File: app/jobs/store_version_contents_job.rb
Comment (simi) [inline app/jobs/store_version_contents_job.rb:4]: What about scope those 2 new jobs into custom queue like `gem_contents`?

File: lib/gem_package_enumerator.rb
Comment (simi) [reply lib/gem_package_enumerator.rb:10]: Merging this back upstream would be the best, but I was about to not block this until that happens and is released.

File: app/jobs/store_version_contents_job.rb
Comment (simi) [inline app/jobs/store_version_contents_job.rb:5]: This is not needed since https://github.com/rubygems/rubygems.org/commit/c791b89bb38a996e773c9a91dc0f93ea9e0560be.

File: app/models/gem_content.rb
Comment (simi) [inline app/models/gem_content.rb:3]: In rubygems.org related codebase package is called `rubygem` and the naming is build around that. There is `Rubygem` model and `rubygem` relations. On the other side, this naming (most likely to represent `Gem::Content` seems using `rubygems` itself namespace `Gem`.

To keep it consistent, I propose to do one of following:

- keep it in `Gem` namespace (thinking about this as an extension to rubygems itself) and move it into `lib`. It is ok to scope it into `Gem::Content`.
- move it into `Rubygem` namespace and keep it into models

File: lib/gem_package_enumerator.rb
Comment (indirect) [reply lib/gem_package_enumerator.rb:10]: I think it makes sense to me to keep the name as-is since this provides the base enumerator as well as (optional) laziness.

Comment (simi) [reply lib/gem_package_enumerator.rb:10]: @indirect Even the name would be the same, is the comment wrong?

Comment (indirect) [reply lib/gem_package_enumerator.rb:10]: Updating the comment to be clear that it adds an enumerator, and provides a lazy enumerator if you do not provide a block, sounds very reasonable to me. 👍🏻

File: .rubocop.yml
Comment (martinemde) [reply .rubocop.yml:13]: Rubocop picks up the indexed content files named "rubocop.yml" and tries to parse them, fails, and then breaks rubocop entirely. We have no need to scan the RubygemFs::Local filesystem for rubocop violations anyway.

File: app/jobs/store_version_contents_job.rb
Comment (martinemde) [reply app/jobs/store_version_contents_job.rb:4]: Good idea.

File: app/jobs/store_version_contents_job.rb
Comment (martinemde) [reply app/jobs/store_version_contents_job.rb:5]: Gotcha. Funny enough an earlier review had me add it. We've been editing this pr for a while.

File: .rubocop.yml
Comment (simi) [reply .rubocop.yml:13]: I have seen that in other PR as well. Anyway it seems relevant actually to this PR. Let's keep it. :muscle:

File: app/models/gem_content.rb
Comment (martinemde) [reply app/models/gem_content.rb:3]: This belongs to Rubygems.org. It's essentially a specialized RubygemFs, so it seems reasonable to change to RubygemContent.

File: app/jobs/store_version_contents_job.rb
Comment (martinemde) [reply app/jobs/store_version_contents_job.rb:31]: It will. As far as I know, we regularly load the entire gem into memory on push, so it's not an unreasonable lift. The way it's processed the max memory usage should be gem + largest file + overhead of stream processing.

Comment (simi) [reply app/jobs/store_version_contents_job.rb:31]: OK, I was thinking about streaming file to temp file first and reading it from disk. I'll take a look later on this. We can do that optionally only for huge files at some point.

File: lib/gem_package_enumerator.rb
Comment (martinemde) [reply lib/gem_package_enumerator.rb:6]: Left the name, cleaned the comment up. Hopefully this goes away if I can upstream the each_entry method, or something like that.

File: app/models/gem_content_entry.rb
Comment (martinemde) [reply app/models/gem_content_entry.rb]: Done.

File: app/jobs/store_version_contents_job.rb
Comment (martinemde) [reply app/jobs/store_version_contents_job.rb]: I used `version_contents` to match the job names.

File: app/jobs/store_version_contents_job.rb
Comment (martinemde) [reply app/jobs/store_version_contents_job.rb]: Fixed

File: app/models/gem_content.rb
Comment (martinemde) [reply app/models/gem_content.rb]: Changed, and I was inspired to cleanup the [plural/singular "contents" inconsistency](https://proofed.com/writing-tips/word-choice-content-vs-contents/). I think it reads better now.

File: app/models/pusher.rb
Comment (segiddins) [inline app/models/pusher.rb:129]: https://app.launchdarkly.com/default/production/features/gemcutter.pusher.store_version_contents/targeting created the flag

File: app/models/version_manifest.rb
Comment (segiddins) [inline app/models/version_manifest.rb:96]: `fs.store spec_key, ruby_spec, content_type: mime`

**Review by segiddins — APPROVED** [review]
did extensive testing on staging. works. let's ship it! (with the FF disabled)

**Comment by simi** [comment]
@martinemde can you squash commits please? I would like to get also one final review round, but overall seems good :muscle:.

**Comment by segiddins** [comment]
@simi we can have GitHub squash when we merge

**Comment by simi** [comment]
@segiddins I'm big fan of GitHub squash button, but for both repos RubyGems/RubyGems.org I have seen merges being done only with merge commits. :thinking: Is it ok to squash or rebase merge as well?

**Comment by segiddins** [comment]
I squash merge everything on this repo

**Comment by simi** [comment]
> I squash merge everything on this repo

Good to know, feel free to ignore my request @martinemde.

**Pull request merged**
